:bug: mitigate upgrade-e2e test flakiness

perdasilva opened the pull request:
# Description

* Mitigates the upgrade-e2e test flakiness caused by #1626 by waiting for the catalog's `last unpacked` to be after the catalogd deployment pod's creation time 
* Updates the upgrade-e2e test to also wait for the catalogd deployment to be up
* Factors out the artifact collection code into a `utils` package and adds artifact collection to upgrade-e2e

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

LalatenduMohanty commented:
We have https://github.com/operator-framework/operator-controller/issues/1550 for this issue as well

azych reviewed (approved):
lgtm

camilamacedo86 commented on .github/workflows/e2e.yaml:
```diff
@@ -58,4 +58,10 @@ jobs:
+      if: failure()
+      with:
+        name: upgrade-e2e-artifacts
+        path: /tmp/artifacts/
```
Thank you 🥇

camilamacedo86 reviewed (approved):
I am OK with the changes
Thank you a lot for this contribution
+1 in the nit raised by @azych 
we might want to address it follow and it across the whole project

/lgtm

LalatenduMohanty commented:
@perdasilva How can I test this PR locally on my laptop? I have local kind setup created usng `make run`.

LalatenduMohanty commented:
I deleted the kind setup and `$ make test-upgrade-e2e` should test this PR.

perdasilva commented:
> @perdasilva How can I test this PR locally on my laptop? I have local kind setup created usng `make run`.

`make test-upgrade-e2e` will exercise the upgrade test

if you want to step through it, it also have the recipe for it:

```
.PHONY: test-upgrade-e2e
test-upgrade-e2e: KIND_CLUSTER_NAME := operator-controller-upgrade-e2e
test-upgrade-e2e: export TEST_CLUSTER_CATALOG_NAME := test-catalog
test-upgrade-e2e: export TEST_CLUSTER_EXTENSION_NAME := test-package
test-upgrade-e2e: kind-cluster run-latest-release image-registry pre-upgrade-setup docker-build kind-load kind-deploy post-upgrade-checks kind-clean #HELP Run upgrade e2e tests on a local kind cluster
```

LalatenduMohanty commented:
I could reproduce the issue without your PR, however when I tested your PR it failed with

```
=== RUN   TestClusterExtensionAfterOLMUpgrade
    post_upgrade_test.go:31: Starting checks after OLM upgrade
    post_upgrade_test.go:36: Wait for catalogd deployment to be ready
    post_upgrade_test.go:121: Checking that the deployment is updated
    post_upgrade_test.go:137: Waiting for only one Pod to remain
    post_upgrade_test.go:40: Wait for operator-controller deployment to be ready
    post_upgrade_test.go:121: Checking that the deployment is updated
    post_upgrade_test.go:137: Waiting for only one Pod to remain
    post_upgrade_test.go:43: Reading logs to make sure that ClusterExtension was reconciled by operator-controller before we update it
    post_upgrade_test.go:55: Checking that the ClusterCatalog is unpacked
    post_upgrade_test.go:56: 
        	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:62
        	            				/usr/local/go/src/runtime/asm_amd64.s:1700
        	Error:      	Expected nil, but got: &v1.Condition{Type:"Serving", Status:"True", ObservedGeneration:1, LastTransitionTime:time.Date(2025, time.January, 16, 7, 42, 20, 0, time.Local), Reason:"Available", Message:"Serving desired content from resolved source"}
    post_upgrade_test.go:56: 
        	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:72
        	            				/usr/local/go/src/runtime/asm_amd64.s:1700
        	Error:      	Expected nil, but got: &v1.Condition{Type:"Progressing", Status:"True", ObservedGeneration:1, LastTransitionTime:time.Date(2025, time.January, 16, 7, 42, 20, 0, time.Local), Reason:"Succeeded", Message:"Successfully unpacked and stored content from resolved source"}
    post_upgrade_test.go:56: 
        	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:56
        	Error:      	Condition never satisfied
        	Test:       	TestClusterExtensionAfterOLMUpgrade
--- FAIL: TestClusterExtensionAfterOLMUpgrade (66.42s)
FAIL
FAIL	github.com/operator-framework/operator-controller/test/upgrade-e2e	66.450s
FAIL
make: *** [Makefile:228: post-upgrade-checks] Error 1
```

perdasilva commented:
> I could reproduce the issue without your PR, however when I tested your PR it failed with
> 
> ```
> === RUN   TestClusterExtensionAfterOLMUpgrade
>     post_upgrade_test.go:31: Starting checks after OLM upgrade
>     post_upgrade_test.go:36: Wait for catalogd deployment to be ready
>     post_upgrade_test.go:121: Checking that the deployment is updated
>     post_upgrade_test.go:137: Waiting for only one Pod to remain
>     post_upgrade_test.go:40: Wait for operator-controller deployment to be ready
>     post_upgrade_test.go:121: Checking that the deployment is updated
>     post_upgrade_test.go:137: Waiting for only one Pod to remain
>     post_upgrade_test.go:43: Reading logs to make sure that ClusterExtension was reconciled by operator-controller before we update it
>     post_upgrade_test.go:55: Checking that the ClusterCatalog is unpacked
>     post_upgrade_test.go:56: 
>         	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:62
>         	            				/usr/local/go/src/runtime/asm_amd64.s:1700
>         	Error:      	Expected nil, but got: &v1.Condition{Type:"Serving", Status:"True", ObservedGeneration:1, LastTransitionTime:time.Date(2025, time.January, 16, 7, 42, 20, 0, time.Local), Reason:"Available", Message:"Serving desired content from resolved source"}
>     post_upgrade_test.go:56: 
>         	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:72
>         	            				/usr/local/go/src/runtime/asm_amd64.s:1700
>         	Error:      	Expected nil, but got: &v1.Condition{Type:"Progressing", Status:"True", ObservedGeneration:1, LastTransitionTime:time.Date(2025, time.January, 16, 7, 42, 20, 0, time.Local), Reason:"Succeeded", Message:"Successfully unpacked and stored content from resolved source"}
>     post_upgrade_test.go:56: 
>         	Error Trace:	/home/lmohanty/code/github.com/OLM-Upstream/operator-controller/test/upgrade-e2e/post_upgrade_test.go:56
>         	Error:      	Condition never satisfied
>         	Test:       	TestClusterExtensionAfterOLMUpgrade
> --- FAIL: TestClusterExtensionAfterOLMUpgrade (66.42s)
> FAIL
> FAIL	github.com/operator-framework/operator-controller/test/upgrade-e2e	66.450s
> FAIL
> make: *** [Makefile:228: post-upgrade-checks] Error 1
> ```

yeah - seems to be due to some recent changes addressing some reviewer comments. Maybe we're finding out why !assert.NotNil was used...?

perdasilva commented on test/upgrade-e2e/post_upgrade_test.go:
```diff
@@ -101,16 +101,47 @@ func TestClusterExtensionAfterOLMUpgrade(t *testing.T) {
 	require.EventuallyWithT(t, func(ct *assert.CollectT) {
 		assert.NoError(ct, c.Get(ctx, types.NamespacedName{Name: testClusterExtensionName}, &clusterExtension))
 		cond := apimeta.FindStatusCondition(clusterExtension.Status.Conditions, ocv1.TypeInstalled)
-		if !assert.NotNil(ct, cond) {
```
this will never return false - either it returns true or fails

LalatenduMohanty replied:
Yup, I agree.

perdasilva commented:
@azych I've changed it to `assert.NotNil` looking at the impl. it seem we either get `true` or a failure. So, I don't think the if made sense to begin with...?

camilamacedo86 commented:
Hi @perdasilva,

I liked the other improvements you made. However, regarding the issue, I wanted to clarify something:

It seems like we are now allocating 1 minute for the load and 1 minute for the update. Wouldn’t that effectively result in a total increase of 2 minutes overall? Or did I misunderstand something?

Looking forward to your thoughts!

perdasilva commented:
> Hi @perdasilva,
> 
> I liked the other improvements you made. However, regarding the issue, I wanted to clarify something:
> 
> It seems like we are now allocating 1 minute for the load and 1 minute for the update. Wouldn’t that effectively result in a total increase of 2 minutes overall? Or did I misunderstand something?
> 
> Looking forward to your thoughts!

Do you mean because we now wait for the catalogd deployment?
Aside from that, which I think is warranted after the mono-repo changes, I don't think there's any change in the time allocation itself. What this change does is ensure that the catalog has been unpacked and is serving before we wait for the update. I think this is still has the same intention as the original test. It's just that the guard for unpack would pass really quickly, since the conditions on the catalog don't change from the installation step. Does that make sense?

LalatenduMohanty reviewed (dismissed):
/lgtm

LalatenduMohanty commented:
/hold for the https://github.com/operator-framework/operator-controller/pull/1627#issuecomment-2595740347 . Feel free to remove the hold.

azych commented:
> @azych I've changed it to `assert.NotNil` looking at the impl. it seem we either get `true` or a failure. So, I don't think the if made sense to begin with...?

👍 I think your current change makes sense looking at the intention

camilamacedo86 commented:
HI @LalatenduMohanty 

Can we unhold this one?
Can we get it merged?

It is 
/lgtm for me !!!

LalatenduMohanty commented:
/hold cancel

camilamacedo86 commented:
/lgtm
